src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -167,41 +183,155 @@ internal int EndRead(IAsyncResult asyncResult)
+
+        internal Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
         {
-            if (asyncResult == null)
+            _sslState.CheckThrow(authSuccessCheck: true, shutdownCheck: true);
+            ValidateParameters(buffer, offset, count);
+
+            if (Interlocked.Exchange(ref _nestedWrite, 1) == 1)
             {
-                throw new ArgumentNullException(nameof(asyncResult));
+                throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, "WriteAsync", "write"));
+            }
+            if (count == 0 && !SslStreamPal.CanEncryptEmptyMessage)
+            {
+                // If it's an empty message and the PAL doesn't support that,
+                // we're done.
+                _nestedWrite = 0;
+                return Task.CompletedTask;
             }
+            if (count <= _sslState.MaxDataSize)
+            {
+                //Single write
+                Task t = WriteSingleChunk(buffer, offset, count, cancellationToken);
+                if (t.IsCompleted)
+                {
+                    _nestedWrite = 0;
 
-            LazyAsyncResult lazyResult = asyncResult as LazyAsyncResult;
-            if (lazyResult == null)
+                    return t;
+                }
+                else
+                {
+                    return ExitWriteAsync(t);
+                }
+            }
+            else
             {
-                throw new ArgumentException(SR.Format(SR.net_io_async_result, asyncResult.GetType().FullName), nameof(asyncResult));
+                //Loop
+                return WriteAsyncChunked(buffer, offset, count, cancellationToken);
             }
 
-            if (Interlocked.Exchange(ref _nestedWrite, 0) == 0)
+            async Task ExitWriteAsync(Task task)
             {
-                throw new InvalidOperationException(SR.Format(SR.net_io_invalidendcall, "EndWrite"));
+                try
+                {
+                    await task.ConfigureAwait(false);
+                }
+                catch (Exception e)
+                {
+                    _sslState.FinishWrite();
+
+                    if (e is IOException)
+                    {
+                        throw;
+                    }
+
+                    throw new IOException(SR.net_io_write, e);
+                }
+                finally
+                {
+                    _nestedWrite = 0;
+                }
             }
+        }
 
-            // No "artificial" timeouts implemented so far, InnerStream controls timeout.
-            lazyResult.InternalWaitForCompletion();
+        private async Task WaitForWriteIOSlot(Task lockTask, byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            await lockTask.ConfigureAwait(false);
+            await WriteSingleChunk(buffer, offset, count, cancellationToken).ConfigureAwait(false);
```

Comment:
Won't this cause us to go do the CheckEnqueueWriteAsync again?

Reply:
The lock is re-entrant , I will double check the logic, but it keeps the flow simple for the "fastpath". If we actually hit this lock it's a 1 in 100 if that. (Only if there is a renegotiation and we try to write). At that point it's slow anyway as we had to wait for a couple of network round trips so a check of an Interlock it won't really hurt out performance.

The reason it's re-entrant because we set it to "LockWrite" when we want to lock and only check that it isn't "LockHandshake".

Reply:
Yeah, I don't think the perf matters, just trying to understand the logic.

We should add a commenting explaining this.

Reply:
:) its a tangled best this lot in general, the more eyes and understanding and the simpler we can make it for those that follow the better!